internal/webhooks/cluster.go

```diff
@@ -366,65 +367,56 @@ func (webhook *Cluster) getClusterClassForCluster(ctx context.Context, cluster *
-			))
+	if cluster.Spec.Topology.ControlPlane.MachineHealthCheck != nil {
+		path := field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck")
+
+		// Validate ControlPlane MachineHealthCheck if defined.
+		if !cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.IsZero() {
+			// Ensure ControlPlane does not define a MachineHealthCheck if the ClusterClass does not define MachineInfrastructure.
+			if clusterClass.Spec.ControlPlane.MachineInfrastructure == nil {
+				allErrs = append(allErrs, field.Forbidden(
+					path,
+					"can be set only if spec.controlPlane.machineInfrastructure is set in ClusterClass",
+				))
+			}
 		}
-		// Ensure ControlPlane MachineHealthCheck defines UnhealthyConditions.
-		if len(cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.UnhealthyConditions) == 0 {
+
+		// If MachineHealthCheck is explicitly enabled then make sure that a MachineHealthCheck definition is
+		// available either in the Cluster topology or in the ClusterClass.
+		// (One of these definitions will be used in the controller to create the MachineHealthCheck)
+		if cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable != nil &&
+			*cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable &&
+			cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.IsZero() &&
+			clusterClass.Spec.ControlPlane.MachineHealthCheck == nil {
 			allErrs = append(allErrs, field.Forbidden(
-				field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck", "unhealthyConditions"),
-				"must have at least one value",
+				path.Child("enable"),
+				fmt.Sprintf("cannot be set to %t as MachineHealthCheck definition is not available in the Cluster topology or the ClusterClass", *cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable),
 			))
 		}
-	}
-
-	// If MachineHealthCheck is explicitly enabled then make sure that a MachineHealthCheck definition is
-	// available either in the Cluster topology or in the ClusterClass.
-	// (One of these definitions will be used in the controller to create the MachineHealthCheck)
-	if cluster.Spec.Topology.ControlPlane.MachineHealthCheck != nil &&
-		cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable != nil &&
-		*cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable &&
-		cluster.Spec.Topology.ControlPlane.MachineHealthCheck.MachineHealthCheckClass.IsZero() &&
-		clusterClass.Spec.ControlPlane.MachineHealthCheck == nil {
-		allErrs = append(allErrs, field.Forbidden(
-			field.NewPath("spec", "topology", "controlPlane", "machineHealthCheck", "enable"),
-			fmt.Sprintf("cannot be set to %t as MachineHealthCheck definition is not available in the Cluster topology or the ClusterClass", *cluster.Spec.Topology.ControlPlane.MachineHealthCheck.Enable),
-		))
+		allErrs = append(allErrs, validateMachineHealthCheckTopology(path, cluster.Namespace, cluster.Spec.Topology.ControlPlane.MachineHealthCheck)...)
```

Comment:
Trickiest section here is trying to correctly merge the MHC Toplogy with the underlying MHC Class to get a validatable result here.